### ISSUE TITLE
Allow excluding class expressions in max-classes-per-file

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [eslint/eslint#11883]
**What rule do you want to change?**

Add an option to `max-classes-per-file` to no longer count class expressions.

Related TSLint issue: https://github.com/palantir/tslint/issues/3270

**Does this change cause the rule to produce more or fewer warnings?**

Fewer.

**How will the change be implemented?

New option. Say, `"allowClassExpressions"`?

**Please provide some example code that this change will affect:**

```js
class VisitorFactory {
    forDescriptor(descriptor) {
        return class {
            visit(node) {
                return `Visiting ${descriptor}.`;
            }
        };
    }
}

// (before editing, the class included a `descriptor = descriptor`: hence the class fields comments
```

**What does the rule currently do for this code?**

Counts both class _declarations_ and _expressions_ no matter what.

**What will the rule do after it's changed?**

If this new option is enabled, only consider class _declarations_.

**Are you willing to submit a pull request to implement this change?**

Absolutely! 🙌

**Comment by g-plane** [comment]
Hey @JoshuaKGoldberg , thanks for the issue.

"Class fields" is still at Stage 3, while ESLint only supports features at Stage 4.

**Comment by kaicataldo** [comment]
Thanks for being willing to make this change. As @g-plane mentions above, our policy is to support features once they reach stage 4, for maintenance reasons. If and when that happens, we would love to accept a PR for this change!

In the meantime, you might be able to solve this using [eslint-rule-composer](https://github.com/not-an-aardvark/eslint-rule-composer) or making an issue over at [eslint-plugin-babel](https://github.com/babel/eslint-plugin-babel).

**Comment by platinumazure** [comment]
I'm confused. The issue request seems to be about not having ClassExpression nodes be counted by this rule. Although the issue creator's example does contain a class field, I don't think that's relevant to the request.

@kaicataldo @g-plane Am I missing something? If not, I think we should reopen this.

**Comment by JoshuaKGoldberg** [comment]
+1 to platinumazure's comments. I'll edit the OP to not include the class field example.

**Comment by kaicataldo** [comment]
Thanks for clarifying. One thing I’m confused about is the option name and it’s behavior as currently proposed here. I wonder if an option name like `includeClassExpressions` would be clearer.

One question I would like to pose is this: Do you think the issue here is that it checks both classes and class expressions or that the rule doesn’t distinguish between scopes? I feel like a change that helps enforce the latter is more in line with the spirit of the rule. I’m not against a change like the one proposed here, but I’m not convinced distinguishing between class declarations and expressions is useful outside of a case like this. Open to being to convinced otherwise, though!
